### DISCUSSION TITLE
TCP transmission from a SED systematically requires retransmission?

### DISCUSSION THREAD

**Discussion by fgervais** [openthread/openthread#7822] [Q&A] [answered]
I have a SED sending data through TCP to a server located through the BR.

My problem is that the SED ends up having to retransmit a lot of TCP packets. Reading this https://github.com/openthread/openthread/issues/5060, it looks like it is the expected behavior but it looks odd and ends up costing more power and time so I'd like some feedback if possible.

Let see what happens just with the SYN/SYN-ACK of the socket opening.

In the following BR log, you can see:

1. SED sends SYN.
2. BR ACK and there is no frames pending (fp: 0)
3. SED sees no frame pending, goes to sleep.
4. 2ms later, BR receives SYN-ACK, puts it in queue.
5. 200ms later, TCP timeouts on SED, SED sends a new SYN.
6. BR ACK with fp:1
7. BR receives SYN-ACK of retry
8. BR receives data poll from SED, replies that it has 2 packets (2 SYN-ACKs)
9. SED moves on a bit with the rest of the transfer until the same will happen a bit later.

```
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: =========================[RX seqno=182 len=069]=========================
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: | 69 98 B6 CD AB 00 08 09 | 08 0D 2A 37 00 00 03 7A | i.6M+.....*7...z
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: | D0 10 06 24 7A 85 C8 82 | 7B 6A DC FD 00 00 64 00 | P..$z.H.{j\}..d.
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: | 00 00 00 00 00 00 00 C0 | A8 02 9F B7 07 07 5B FB | .......@(..7..[{
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: | 96 40 B7 00 00 00 00 50 | 02 05 00 5A B7 00 00 DE | .@7....P...Z7..^
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: | 4B A4 70 28 B0 .. .. .. | .. .. .. .. .. .. .. .. | K$p(0...........
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.558 [DEBG]-MAC-----: ------------------------------------------------------------------------
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.559 [INFO]-MAC-----: Received IPv6 TCP msg, len:60, chksum:5ab7, from:0x0809, sec:yes, prio:normal, rss:-52.0
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.559 [INFO]-MAC-----:     src:[fd11:22:0:0:247a:85c8:827b:6adc]:46855
Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.559 [INFO]-MAC-----:     dst:[fd00:64:0:0:0:0:c0a8:29f]:1883

Jun 18 14:12:08 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.561 [INFO]-MAC-----: Prepping indir tx IPv6 TCP msg, len:64, chksum:cbd4, to:0x0809, sec:yes, prio:low

Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: =========================[RX seqno=183 len=069]=========================
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: | 69 98 B7 CD AB 00 08 09 | 08 0D 2B 37 00 00 03 7A | i.7M+.....+7...z
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: | D0 10 06 24 7A 85 C8 82 | 7B 6A DC FD 00 00 64 00 | P..$z.H.{j\}..d.
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: | 00 00 00 00 00 00 00 C0 | A8 02 9F B7 07 07 5B FB | .......@(..7..[{
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: | 96 40 B7 00 00 00 00 50 | 02 05 00 5A B7 00 00 3A | .@7....P...Z7..:
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: | EB 72 6C 27 4E .. .. .. | .. .. .. .. .. .. .. .. | krl'N...........
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [DEBG]-MAC-----: ------------------------------------------------------------------------
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [INFO]-MAC-----: Received IPv6 TCP msg, len:60, chksum:5ab7, from:0x0809, sec:yes, prio:normal, rss:-53.0
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [INFO]-MAC-----:     src:[fd11:22:0:0:247a:85c8:827b:6adc]:46855
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.758 [INFO]-MAC-----:     dst:[fd00:64:0:0:0:0:c0a8:29f]:1883


Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.761 [INFO]-MAC-----: Rx data poll, src:0x0809, qed_msgs:2, rss:-52, ack-fp:1


Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: =========================[TX seqno=043 len=074]=========================
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: | 79 98 2B CD AB 09 08 00 | 08 0D D0 10 00 00 03 78 | y.+M+.....P....x
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: | 85 01 06 3B FD 00 00 64 | 00 00 00 00 00 00 00 00 | ...;}..d........
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: | C0 A8 02 9F 24 7A 85 C8 | 82 7B 6A DC 07 5B B7 07 | @(..$z.H.{j\.[7.
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: | 52 44 2E E0 FB 96 40 B8 | 60 12 FA F0 CB D4 00 00 | RD.`{.@8`.zpKT..
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: | 02 04 05 B4 DD 37 3A 13 | 5A AC .. .. .. .. .. .. | ...4]7:.Z,......
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.767 [DEBG]-MAC-----: ------------------------------------------------------------------------
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.768 [INFO]-MAC-----: Sent IPv6 TCP msg, len:64, chksum:cbd4, to:0x0809, sec:yes, prio:low
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.768 [INFO]-MAC-----:     src:[fd00:64:0:0:0:0:c0a8:29f]:1883
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.768 [INFO]-MAC-----:     dst:[fd11:22:0:0:247a:85c8:827b:6adc]:46855

Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.768 [INFO]-MAC-----: Prepping indir tx IPv6 TCP msg, len:64, chksum:cbd4, to:0x0809, sec:yes, prio:low

Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.774 [INFO]-MAC-----: Rx data poll, src:0x0809, qed_msgs:1, rss:-52, ack-fp:1
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: =========================[TX seqno=044 len=074]=========================
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: | 69 98 2C CD AB 09 08 00 | 08 0D D1 10 00 00 03 78 | i.,M+.....Q....x
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: | 85 01 06 3B FD 00 00 64 | 00 00 00 00 00 00 00 00 | ...;}..d........
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: | C0 A8 02 9F 24 7A 85 C8 | 82 7B 6A DC 07 5B B7 07 | @(..$z.H.{j\.[7.
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: | 52 44 2E E0 FB 96 40 B8 | 60 12 FA F0 CB D4 00 00 | RD.`{.@8`.zpKT..
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: | 02 04 05 B4 DD 37 3A 13 | 5A AC .. .. .. .. .. .. | ...4]7:.Z,......
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [DEBG]-MAC-----: ------------------------------------------------------------------------
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [INFO]-MAC-----: Sent IPv6 TCP msg, len:64, chksum:cbd4, to:0x0809, sec:yes, prio:low
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [INFO]-MAC-----:     src:[fd00:64:0:0:0:0:c0a8:29f]:1883
Jun 18 14:12:09 3b31dc0359e9 otbr-agent[176]: 6d.22:56:50.780 [INFO]-MAC-----:     dst:[fd11:22:0:0:247a:85c8:827b:6adc]:46855

```

Here you can see what is looks like on the server side with all the retransmissions:

![image](https://user-images.githubusercontent.com/5411996/174462741-02761c4f-106d-4fc9-93b5-518118f637d8.png)

Is there a way around this?

I feel, at least for my use-case, it is much better to just set this to always `true`:

https://github.com/openthread/openthread/blob/481a064f034496d93273f4b837d46559f7b74e5c/src/core/mac/data_poll_sender.cpp#L361

**Comment by fgervais** [comment]
I kept working on this even though I'm actually not sure if I'm just reinventing the wheel or actually solving a real issue.

My previous hack ended up unreliable as sometimes the data poll would actually happen too fast, before the BR had time to get the SYN-ACK and so I would gain nothing.

However I found out the concept of the "Fast Poll" and this seems exactly the kind of thing I need.

I checked in the openthread tcp stack and it doesn't seem like it does enable fast polling when expecting a tcp reply but I haven't researched that a ton. On my side, I'm using the zephyr tcp stack and it sure doesn't.

I think what would need to happen is for the tcp stack to toggle between normal and fast polling depending of the state of the socket. 

However, as a quick proof of concept I enabled it all the time:

```patch
diff --git a/src/core/mac/data_poll_sender.cpp b/src/core/mac/data_poll_sender.cpp
index ab1ad0db6..622bb59bc 100644
--- a/src/core/mac/data_poll_sender.cpp
+++ b/src/core/mac/data_poll_sender.cpp
@@ -377,6 +377,10 @@ void DataPollSender::ProcessTxDone(const Mac::TxFrame &aFrame, const Mac::RxFram
     {
         IgnoreError(SendDataPoll());
     }
+    else
+    {
+        SendFastPolls(1);
+    }
 
 exit:
     return;
diff --git a/src/core/mac/data_poll_sender.hpp b/src/core/mac/data_poll_sender.hpp
index a30f56362..d3ad678e6 100644
--- a/src/core/mac/data_poll_sender.hpp
+++ b/src/core/mac/data_poll_sender.hpp
@@ -269,7 +269,7 @@ private:
     // Poll period under different conditions (in milliseconds).
     static constexpr uint32_t kAttachDataPollPeriod = OPENTHREAD_CONFIG_MAC_ATTACH_DATA_POLL_PERIOD;
     static constexpr uint32_t kRetxPollPeriod       = OPENTHREAD_CONFIG_MAC_RETX_POLL_PERIOD;
-    static constexpr uint32_t kFastPollPeriod       = 188;
+    static constexpr uint32_t kFastPollPeriod       = 10;
     static constexpr uint32_t kMinPollPeriod        = OPENTHREAD_CONFIG_MAC_MINIMUM_POLL_PERIOD;
     static constexpr uint32_t kMaxExternalPeriod    = ((1 << 26) - 1); // ~18.6 hours.

```

With this my transaction with the MQTT server finishes in ~70ms instead of the ~650ms you see in the screenshot of my first post:

![image](https://user-images.githubusercontent.com/5411996/174847908-a117f99e-df65-416a-bd8d-196b1c62a86e.png)

This is really the difference between instant and sluggish from a humain point of view.

Any idea on this?

I'm new to this spec so I might well be trying to solve my problem the wrong way.

**Comment by jwhui** [comment, marked as answer]
Thanks for raising this issue. You are correct that we have not done the work to integrate TCP with Data Polling on a Sleepy End Device (SED). I think we'll have to do some more thinking around how best to integrate data poll period adjustments with TCP.

While we could have the SED poll more frequently while it is waiting for a TCP ACK, it's not clear how the SED would determine when to poll more frequently when the peer is sending data to it. The challenge is that the SED does not really know when it might receive data from a peer.

One thing you could try is by using [`otLinkSetPollPeriod()`](https://openthread.io/reference/group/api-link-link#otlinksetpollperiod) to set the poll period. Of course, setting the poll period to something small will increase the SEDs power draw. While it would be nice to have this more integrated within the TCP stack itself, the application typically knows better when it might receive data from the peer.

**Comment by fgervais** [comment]
I gave the `otLinkSetPollPeriod` route a try and it does work.

I do a `otLinkSetPollPeriod(instance, 10);` just before starting the connection and a `otLinkSetPollPeriod(instance, 0);` after close.

One thing I would point out is that it's not trivial for the application to know exactly when to stop the aggressive polling.

This is application specific but at least with zephyr `v3.0.99`, if I stop the aggressive polling when I receive my callback that I disconnected from the MQTT server, it is still too early and there will be some retransmissions in the socket termination if I stop aggressive polling right away. I need to wait a bit more time to make sure like 50ms or something.

For the moment, on my side, this is good enough, thank you very much.
